src/Data/AIG/CompactGraph.hs

```diff
@@ -0,0 +1,444 @@
+-- | Turn a variable into its positive literal.
+varToLit :: Var -> CompactLit s
+varToLit (Var v) = CompactLit (v `shiftL` 1)
+
+-- | Extract the variable a literal refers to.
+litToVar :: CompactLit s -> Var
+litToVar (CompactLit l) = Var (l `shiftR` 1)
+
+-- | Determine whether the given literal is negated.
+litNegated :: CompactLit s -> Bool
+litNegated (CompactLit l) = testBit l 0
+
+-- | Return the second literal with its polarity altered to match the
+-- first.
+copySign :: CompactLit s -> CompactLit s -> CompactLit s
+copySign (CompactLit src) (CompactLit dst) =
+  CompactLit ((dst .&. 0xFFFFFFFE) .|. (src .&. 0x00000001))
+
+newCompactGraph :: IO (CompactGraph s)
+newCompactGraph =
+  do maxVar  <- newIORef (Var 0)
+     inputs  <- newIORef []
+     andMap  <- newIORef Map.empty
+     return (CompactGraph {..})
+
+newVar :: CompactGraph s -> IO Var
+newVar g =
+  do modifyIORef' (maxVar g) nextVar
+     readIORef (maxVar g)
+
+-- | Create a map associating a new destination variable with each
+-- existing variable. Used to ensure the ordering invariants of the
+-- binary AIGER file format.
+mkVarMap ::
+  [Var] ->
+  Map Var (CompactLit s, CompactLit s) ->
+  (Map Var Var)
+mkVarMap ins gateMap =
+  Map.fromList (zip varList [Var 0..])
+    where
+      varList = [Var 0] ++ ins ++ Map.keys gateMap
+
+-- | Adjust a literal according to the given variable mapping.
+lookupLit :: CompactLit s -> Map Var Var -> Maybe (CompactLit s)
+lookupLit l m = (copySign l . varToLit) <$> Map.lookup (litToVar l) m
+
+------------------------------------------------------------------
+-- AIG format
+
+-- | The type of AIGER file to create. We don't currently support user
+-- specification of this. The distinction primarily exists to ease
+-- debugging.
+data AIGFileMode
+  = ASCII
+  | Binary
+
+modeString :: AIGFileMode -> BS.ByteString
+modeString ASCII = "aag"
+modeString Binary = "aig"
+
+------------------------------------------------------------------
+-- AIG generation
+
+bsUnwords :: [BS.Builder] -> BS.Builder
+bsUnwords = mconcat . intersperse (BS.charUtf8 ' ')
+
+hPutBBLn :: Handle -> BS.Builder -> IO ()
+hPutBBLn h b = BS.hPutBuilder h (b <> BS.charUtf8 '\n')
+
+-- | Write an AIGER header line to the given handle.
+writeHeader ::
+  Handle ->
+  AIGFileMode ->
+  Var ->
+  [Var] ->
+  Int ->
+  [CompactLit s] ->
+  Map Var (CompactLit s, CompactLit s) ->
+  IO ()
+writeHeader h format (Var var) ins latches outs gateMap =
+  do hPutBBLn h $ bsUnwords [ BS.byteString (modeString format)
+                            , BS.word32Dec var
+                            , BS.intDec (length ins)
+                            , BS.intDec latches
+                            , BS.intDec (length outs)
+                            , BS.intDec (Map.size gateMap)
+                            ]
+
+-- | Write AIGER input lines to the given handle.
+writeInputs :: Handle -> AIGFileMode -> Int -> Map Var Var -> [Var] -> IO ()
+writeInputs _ Binary _ _ _ = return ()
```

Comment:
If we don't handle this case, we should probably fail loudly here instead of just returning nil.  If there is actually nothing to do here, let's add a comment to that effect.

Reply:
Ah, yeah, the binary mode doesn't include input lists because they can be inferred. I added a comment about that.